[Streams] YAML mode

Kerry350 opened the pull request:
## Summary 

Closes https://github.com/elastic/streams-program/issues/446

This introduces YAML mode to the Streams' processing UI, this means users can now write and edit Streamlang directly.

## Functionality
<details>
<summary>Example images</summary>

<img width="1386" height="574" alt="Screenshot 2025-11-16 at 22 21 22" src="https://github.com/user-attachments/assets/7cd19065-30d2-42e7-b1c1-4367f197af2a" />

<img width="1390" height="820" alt="Screenshot 2025-11-16 at 22 20 29" src="https://github.com/user-attachments/assets/d74d0b60-5ee8-4b03-a2c1-f7ac230301b3" />

<img width="1410" height="954" alt="Screenshot 2025-11-16 at 22 20 45" src="https://github.com/user-attachments/assets/835fc59d-139a-465c-8e64-fcfe7c601805" />

<img width="1342" height="536" alt="Screenshot 2025-11-16 at 22 20 50" src="https://github.com/user-attachments/assets/e517f28a-e4d1-48f8-b235-f02157458f29" />

<img width="1232" height="686" alt="Screenshot 2025-11-16 at 22 21 11" src="https://github.com/user-attachments/assets/eb406c96-4c75-43a3-92a5-a313d2c5b107" />

<img width="1552" height="494" alt="Screenshot 2025-11-16 at 22 18 37" src="https://github.com/user-attachments/assets/fb5e616e-2780-40ec-ae5d-398f76b2e872" />

<img width="824" height="344" alt="Screenshot 2025-11-16 at 22 18 47" src="https://github.com/user-attachments/assets/50682726-3b88-49c0-baae-b4f3bc54eef7" />

<img width="1362" height="528" alt="Screenshot 2025-11-16 at 22 19 52" src="https://github.com/user-attachments/assets/bc94505a-2cc6-4ecb-b7c7-8212fcd182d2" />

<img width="1346" height="784" alt="Screenshot 2025-11-16 at 22 20 03" src="https://github.com/user-attachments/assets/7c646293-6cb3-4cd3-bf95-d4897c0bc9df" />

<img width="948" height="324" alt="Screenshot 2025-11-16 at 23 23 53" src="https://github.com/user-attachments/assets/6a588554-b89d-4d6d-8427-dc9740088449" />

</details>

## Reviewer notes

*There are also some comments inline with the code*.

- Added a `@kbn/streamlang-yaml-editor` package, a wrapper around Monaco that knows how to represent Streamlang. Native Monaco functionality has been used (e.g. decorations, glyph markers etc) to represent things like simulation states, metrics information etc. 

- Introduced a `validateStreamlang()` function in the Streamlang package rather than having Streamlang validation logic spread out in the Streams plugin.

- Added diffing utilities to the Streamlang package to be able to gather information between two sets of DSL.

- The state machines have been refactored: the top level state machine now spawns a child interactive / YAML mode state machine to handle the modes independently since their logic is quite different.
  - The mode is not persisted in the URL at the moment (we can monitor this for requests).
  - At first the `evaluatingMode` state will be entered to ascertain whether we need interactive mode or YAML mode, by default we'll use interactive mode **unless** the configuration can't be represented in interactive mode (e.g. nesting level is too deep).
  - From a very high-level the mode machines track state in their own way, and then sync this (the DSL) back to the top level machine. 

- Simulation in YAML mode **mostly** matches interactive mode with a few differences:
  - Simulation is explicit (using a button)
  - You can either run a full simulation or "run up to step" 
    - "Full" means all steps with complete data sources, and all additive steps with partial data sources.
    - "Run up to step" will either run all steps up to that breakpoint with complete data sources, or all additive steps up to that step with partial data sources.
    - "Run up to step" mirrors the behaviour you'd get in interactive mode when editing a processor, that processor in draft mode acts as the breakpoint for the steps.

- This PR also includes the refactor and migration from `where` to `condition` due to https://github.com/elastic/streams-program/discussions/646, this is relatively self contained in this commit: https://github.com/elastic/kibana/pull/242743/commits/3ddf6e6d56fdc6980bd7ced561f0463916e38761

## Followups 

- It's been a PITA getting Scout to work with the editor nicely (in the sense that I can focus it and move the caret but it doesn't trigger the same things programatically that it does when used "for real"), so I'm going to do the new Scout tests in a followup.

- The enhanced Grok highlighting for the preview table is currently disabled in YAML mode due to the fact we need to deal with the "resources" and grok UI models. What I'd like to do is do a small refactor at the Grok UI package level so that we can actually do away with those resources in the state machine entirely. I'll push that to a followup issue though.

## Open questions

- Should we force an active simulation before being able to save in YAML mode? This is purely because of this logic for the management bar:

```ts
 onConfirm={
            detectedFields.length > 0 && getChanges(detectedFields, definitionFields).length > 0
              ? openConfirmationModal
              : saveChanges
          }
```

we won't get the review modal if there isn't a simulation result available. It's not a bug per se, just a question about behaviour.

Kerry350 committed:
YAML mode

Adds a comprehensive editing experience for Streamlang in YAML mode.

kibanamachine committed:
Changes from node scripts/lint_packages --fix

kibanamachine committed:
Changes from node scripts/generate codeowners

Kerry350 committed:
Amend CODEOWNERS

Kerry350 commented on x-pack/platform/packages/shared/kbn-streamlang/src/schema/get_json_schema_from_streamlang_schema.ts:
```diff
@@ -0,0 +1,501 @@
+ * Monaco cannot resolve, and it omits strict `additionalProperties` flags.
+ * This helper normalises those quirks before returning the schema artifact.
+ */
+export function getJsonSchemaFromStreamlangSchema(
```
Quite a lot of ceremony in here because we don't have a discriminator on our step types (between `action` vs `where` steps), which trips up the JSON schema. Then some of the additional functionality is so that we get a nice display in the Monaco menu.

Kerry350 committed:
Small tweaks, i18n etc

kibanamachine committed:
Changes from node scripts/generate codeowners

Kerry350 committed:
Remove unit tests (as they live in a different location), add unit tests for Streamlang validations.

Kerry350 committed:
Merge branch '446-basic-yaml-editor' of github.com:Kerry350/kibana into 446-basic-yaml-editor

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/361963)
* Commit: f9ea614b4fc61946ab736e6f01de1d511acea73f

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/07655a0a20808401ae4531541ed85885b2f61fea) | [after](https://github.com/elastic/kibana/commit/f9ea614b4fc61946ab736e6f01de1d511acea73f) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 805 | 951 | +146 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/07655a0a20808401ae4531541ed85885b2f61fea) | [after](https://github.com/elastic/kibana/commit/f9ea614b4fc61946ab736e6f01de1d511acea73f) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 216 | 297 | +81 |
| `@kbn/streamlang-yaml-editor` | - | 34 | +34 |
| total |  |  | +115 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/07655a0a20808401ae4531541ed85885b2f61fea) | [after](https://github.com/elastic/kibana/commit/f9ea614b4fc61946ab736e6f01de1d511acea73f) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 941.6KB | 1.1MB | :warning: +208.2KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/07655a0a20808401ae4531541ed85885b2f61fea) | [after](https://github.com/elastic/kibana/commit/f9ea614b4fc61946ab736e6f01de1d511acea73f) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 5 | 6 | +1 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/07655a0a20808401ae4531541ed85885b2f61fea) | [after](https://github.com/elastic/kibana/commit/f9ea614b4fc61946ab736e6f01de1d511acea73f) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 17.1KB | 17.2KB | +72.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/07655a0a20808401ae4531541ed85885b2f61fea) | [after](https://github.com/elastic/kibana/commit/f9ea614b4fc61946ab736e6f01de1d511acea73f) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang` | 233 | 323 | +90 |
| `@kbn/streamlang-yaml-editor` | - | 46 | +46 |
| total |  |  | +136 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/07655a0a20808401ae4531541ed85885b2f61fea) | [after](https://github.com/elastic/kibana/commit/f9ea614b4fc61946ab736e6f01de1d511acea73f) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang-yaml-editor` | - | 7 | +7 |
| `streamsApp` | 17 | 16 | -1 |
| total |  |  | +6 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/07655a0a20808401ae4531541ed85885b2f61fea) | [after](https://github.com/elastic/kibana/commit/f9ea614b4fc61946ab736e6f01de1d511acea73f) | diff |
| --- | --- | --- | --- |
| `@kbn/streamlang-yaml-editor` | - | 7 | +7 |
| `streamsApp` | 27 | 26 | -1 |
| total |  |  | +6 |


</details>


### History
* :broken_heart: [Build #361662](https://buildkite.com/elastic/kibana-pull-request/builds/361662) failed 989eedda8ab87a6a5f3f24415e75b28e6fff14d2
* :broken_heart: [Build #361649](https://buildkite.com/elastic/kibana-pull-request/builds/361649) failed a35cbaea0e6c9bf93f6ed852e06669053001c638
* :broken_heart: [Build #361354](https://buildkite.com/elastic/kibana-pull-request/builds/361354) failed e7ba737fb98a0052ce2f6aa5511ad5b62f3920cc

cc @Kerry350

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361963","number":361963,"commit":"f9ea614b4fc61946ab736e6f01de1d511acea73f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361662","number":361662,"commit":"989eedda8ab87a6a5f3f24415e75b28e6fff14d2"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361649","number":361649,"commit":"a35cbaea0e6c9bf93f6ed852e06669053001c638"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361354","number":361354,"commit":"e7ba737fb98a0052ce2f6aa5511ad5b62f3920cc"}],"number":361963}
buildkite-pr-comment-->

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Move validation to new home

flash1293 reviewed (commented):
This is really cool! Noted down a couple things I noticed in the behavior

The green background makes selection disappear (I selected row 5-9 in this screenshot):

<img width="601" height="264" alt="Image" src="https://github.com/user-attachments/assets/dfa005a1-cdfe-4110-bee0-84b650ff6d2f" />


Seems like the annotations in the editor are not refreshed properly, depending on the action:

![Image](https://github.com/user-attachments/assets/0bb45300-3293-4a82-99d4-40d56c751d27)

Can happen later, but I was really looking for a shortcut like cmd+Enter to submit a step. Right now I always have to get off the keyboard which is pretty disrupting.

Maybe we can avoid adding stuff like
```
    ignore_failure: true
    ignore_missing: true
```

as long as it's the default value when switching to yaml mode? This would make it quite a bit cleaner.


When within a `where` block, the suggestions go a bit crazy:
<img width="445" height="312" alt="Image" src="https://github.com/user-attachments/assets/6b218187-5d83-4c3b-a9af-c40de358e4d5" />

Another case where it gets confused a bit:

<img width="488" height="220" alt="Image" src="https://github.com/user-attachments/assets/091fcbfc-3795-4854-9290-247e0a852a79" />

Current code is:
```
steps:
  - action: grok
    where:
      always: {}
    patterns:
      - "%{WORD:attributes.abc}"
    from: stream.name
    ignore_failure: true
    ignore_missing: true
  - where:
      field: attributes.ip
      eq: 8.8.8.8
      steps:
```

Judging from the warnings, it evaluates each step as a grok step.


Manual ingest pipeline is shown in wired streams:
<img width="549" height="157" alt="Image" src="https://github.com/user-attachments/assets/1ba8cf34-0b47-4e20-92a2-392d51e9c6be" />


In case the streamlang is invalid, it shows a good error in the editor, but the bottom error callout just always shows:
<img width="589" height="99" alt="Image" src="https://github.com/user-attachments/assets/0c6c2bfc-39af-4054-b346-5051f15a4af3" />


which isn't very helpful.

Also, should we use the "streamlang" name user-facing? Or something else? Not sure whether we discussed this, cc @LucaWintergerst 



Switching between the modes makes the bottom bar bounce:

![Image](https://github.com/user-attachments/assets/170718bc-5dcd-4f0c-b25b-8d50438d9b82)

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Transparentize deocration background colours

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Migrate where > condition property for where blocks

Kerry350 committed:
Schema tweaks now we have better discrimination

Kerry350 committed:
Hide manual_ingest_pipeline action from Wired streams in the editor

Kerry350 committed:
Show more detailed errors

Kerry350 committed:
Add keyboard shortcut for running simulation

Kerry350 committed:
Improve monaco-yaml registration (to stop any worker disposal errors)

Kerry350 committed:
Simplify debounce logic and maintain initial simulation result rendering

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Just disable button if invalid

Kerry350 committed:
Expand errors and remove console warning logs

Kerry350 committed:
Fix test and type check

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

Kerry350 committed:
Stabilise bottom bar

Kerry350 committed:
Merge branch '446-basic-yaml-editor' of github.com:Kerry350/kibana into 446-basic-yaml-editor

Kerry350 commented:
@flash1293 This should be ready for another look.

> The green background makes selection disappear

This is improved but not exactly the same as Workflows. TL;DR is the Sass variable resolves to two different colours even though it's the same class in our top level shared stylesheets. I believe this is because Workflows do theme overrides. It's definitely visible now though, so I'll try and address fully in a followup if not good enough.

<img width="1726" height="715" alt="Screenshot 2025-11-20 at 14 15 41" src="https://github.com/user-attachments/assets/9919e94f-16dc-4415-83b1-cd0714d8d2f2" />

<img width="1728" height="690" alt="Screenshot 2025-11-20 at 14 16 03" src="https://github.com/user-attachments/assets/1d4c1719-bd9e-4cbf-beaf-f3735b10060d" />


> Seems like the annotations in the editor are not refreshed properly, depending on the action

Fixed.

> Can happen later, but I was really looking for a shortcut like cmd+Enter to submit a step. Right now I always have to get off the keyboard which is pretty disrupting.

Cmd + Enter will now trigger a simulation run.

> When within a where block, the suggestions go a bit crazy
> Another case where it gets confused a bit

Everything schema related should be fixed now with the `where` to `condition` migration.

> Manual ingest pipeline is shown in wired streams

This should be fixed.

> In case the streamlang is invalid, it shows a good error in the editor, but the bottom error callout just always shows

This is fixed, better errors are shown. **Please note that this is obscured by the layout bug we have currently**.

> Also, should we use the "streamlang" name user-facing? Or something else? Not sure whether we discussed this.

Confirmed okay by Luca.

> Switching between the modes makes the bottom bar bounce

Changed the logic here.

Because of the surface area this PR is a bit of a nightmare for merge conflicts. Please could we separate changes into "critical" and "followup". There's definitely some non-blocking refinements I can think of already, but I'd like to get it merged ASAP if possible once the blockers are out of the way.

kibanamachine committed:
Changes from node scripts/capture_oas_snapshot   --include-path /api/status   --include-path /api/alerting/rule/   --include-path /api/alerting/rules   --include-path /api/actions   --include-path /api/security/role   --include-path /api/spaces   --include-path /api/streams   --include-path /api/fleet   --include-path /api/saved_objects/_import   --include-path /api/saved_objects/_export   --include-path /api/maintenance_window   --include-path /api/agent_builder --update

kibanamachine committed:
Changes from make api-docs

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/367847)
* Commit: 89a518ef80ab0aab21a531bcfc6196cd96bb3ac2

### Failed CI Steps
* [:pipeline::magic_wand: Run Dynamic Pipeline: .buildkite/scripts/pipelines/pull_request/pipeline.sh](https://buildkite.com/elastic/kibana-pull-request/builds/367847#019adfff-679f-40eb-9279-3ef8dc76455b)
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/367847#019ae10a-2b6d-40d7-a6df-d217c9f31680)

### History
* :yellow_heart: [Build #362292](https://buildkite.com/elastic/kibana-pull-request/builds/362292) was flaky e396544cffc6662450b624e064a846fbe9333860
* :broken_heart: [Build #362283](https://buildkite.com/elastic/kibana-pull-request/builds/362283) failed 0c340fab507b7571a9cdb55b2bcefc9ad548427b
* :green_heart: [Build #361963](https://buildkite.com/elastic/kibana-pull-request/builds/361963) succeeded f9ea614b4fc61946ab736e6f01de1d511acea73f
* :broken_heart: [Build #361662](https://buildkite.com/elastic/kibana-pull-request/builds/361662) failed 989eedda8ab87a6a5f3f24415e75b28e6fff14d2
* :broken_heart: [Build #361649](https://buildkite.com/elastic/kibana-pull-request/builds/361649) failed a35cbaea0e6c9bf93f6ed852e06669053001c638
* :broken_heart: [Build #361354](https://buildkite.com/elastic/kibana-pull-request/builds/361354) failed e7ba737fb98a0052ce2f6aa5511ad5b62f3920cc

cc @Kerry350

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367847","number":367847,"commit":"89a518ef80ab0aab21a531bcfc6196cd96bb3ac2"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362292","number":362292,"commit":"e396544cffc6662450b624e064a846fbe9333860"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362283","number":362283,"commit":"0c340fab507b7571a9cdb55b2bcefc9ad548427b"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361963","number":361963,"commit":"f9ea614b4fc61946ab736e6f01de1d511acea73f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361662","number":361662,"commit":"989eedda8ab87a6a5f3f24415e75b28e6fff14d2"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361649","number":361649,"commit":"a35cbaea0e6c9bf93f6ed852e06669053001c638"}],"number":367847}
buildkite-pr-comment-->

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Merge branch '446-basic-yaml-editor' of github.com:Kerry350/kibana into 446-basic-yaml-editor

Kerry350 committed:
Update codeowners

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Tweak

Kerry350 committed:
Amend tweak -_-

Kerry350 committed:
Accomodate bottom bar

flash1293 commented:
Not a blocker - Wrong error message is shown when YAML is invalid (all steps are new here):

<img width="705" height="602" alt="Screenshot 2025-12-04 at 11 33 43" src="https://github.com/user-attachments/assets/4027a651-8ef7-4b34-981d-899ba8f1681b" />

flash1293 commented:
Somehow the context menu is rendered in the wrong place in dark mode - I first thought it's a race condition, but it's reproducable:

<img width="627" height="591" alt="Screenshot 2025-12-04 at 11 46 16" src="https://github.com/user-attachments/assets/ad7abfcf-f00e-4d9b-9fe0-e6e51a9e5983" />

flash1293 commented on x-pack/platform/packages/shared/kbn-streamlang/src/actions/action_metadata.ts:
```diff
@@ -0,0 +1,362 @@
+  tips?: string[];
+}
+
+export const ACTION_METADATA_MAP: Record<string, ActionMetadata> = {
```
How do we keep these up to date? It's already missing `drop_document` and `replace`, I guess that will stay a theme.

<img width="289" height="326" alt="Image" src="https://github.com/user-attachments/assets/f5541123-8fbf-4043-a853-c3d8b6133d05" />

Maybe we can do a switch exhaustiveness check or something like that. Ideally it's all centralized in a manually managed OAS (cc @tonyghiani )

flash1293 commented on x-pack/platform/packages/shared/kbn-streamlang/src/validation/validations.ts:
```diff
@@ -0,0 +1,260 @@
+  });
+}
+
+export const validateStreamlang = (dsl: StreamlangDSL, streamType?: StreamType): true => {
```
Your `validateStreamlang` is pretty close to mine in terms of how it's integrated, they should fit right together.

flash1293 commented:
Side note, but we need to debounce the custom documents source editor - it currently passes through every single keypress which just doesn't fly with all the work we are doing:

<img width="505" height="393" alt="Screenshot 2025-12-04 at 12 14 36" src="https://github.com/user-attachments/assets/74b01905-14f5-46aa-a3ba-a45b0e06cbe5" />

We could even reduce that to just update when you close the flyout or something

flash1293 commented on x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/streams/classic_stream.ts:
```diff
@@ -221,6 +223,21 @@ export class ClassicStream extends StreamActiveRecord<Streams.ClassicStream.Defi
 
+    if (this._changes.processing && this._definition.ingest.processing.steps.length > 0) {
+      try {
+        validateStreamlang(this._definition.ingest.processing, 'classic');
```
Should we do this? On my validation PR I didn't validate this kind of stuff for classic streams because I didn't want to break existing usages and most of the stuff we validate is only relevant for wired streams anyway

flash1293 reviewed (commented):
Some comments for now, will continue review

Kerry350 commented on x-pack/platform/packages/shared/kbn-streamlang/src/utilities/diff_utilities.ts:
```diff
@@ -0,0 +1,269 @@
+  }
+
+  const stepPath = `${path}.steps[${index}]`;
+  const contentHash = objectHash(step);
```
The full step path is probably enough. This was originally a lot more complex as I'd implemented this to try and diff changes that weren't just additive, then we made the decision to go additive only which simplified things.

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Merge resolution: refactors state machines etc to work with "suggest a pipeline" functionality

Kerry350 committed:
Merge resolution: change to work with "outdated documents"

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Merge resolution: integrates with the new validation functionality

Kerry350 committed:
Improve IDs for the individual step machines

Kerry350 commented:
⚠️ ℹ️ Just a heads up that I've pushed several commits that deal with the merge conflicts from several features.

This commit: https://github.com/elastic/kibana/pull/242743/commits/0730c0afa0a6d63a1c8174f7ced7ca18ac9c38ea integrates the "suggest pipelines" functionality with the new state machine setup etc. It introduces a parallel state in the interactive state machine.

This commit: https://github.com/elastic/kibana/pull/242743/commits/ca12df3eaee0fb8d1e74be3452bc3f2e83bc002d integrates with the new validation functionality. The validation errors are shown as gutter markers, and at the bottom in YAML mode.

<img width="721" height="649" alt="Screenshot 2025-12-05 at 03 08 52" src="https://github.com/user-attachments/assets/cfe77ae0-760e-4303-8056-04ee9772bbd6" />

⚠️ There is a bug I'm aware of where if you edit something in interactive mode (causing an error) then switch to YAML mode it doesn't initially show the error, this is due to the editor reference not causing `useEffect` to trigger in `useGutterValidationMarkers` (if you make a further edit, this will work).

⚠️ Also, now that "warn about outdated docs" is in I can see we're rendering `updated_at` in the editor, I'll also fix that.

It's very late here so I'll carry on with everything tomorrow (well, today 😅).

tonyghiani commented on x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/state_management/simulation_state_machine/simulation_state_machine.ts:
```diff
@@ -173,20 +173,14 @@ export const simulationMachine = setup({
         },
-        target: '.assertingRequirements',
         actions: [{ type: 'storeSteps', params: ({ event }) => event }],
+        target: '.debouncingChanges',
```
Is this required under what scenario? The difference in targeting directly the `.assertingRequirements` state is that for some immediate interaction (like click on a button, etc.) we don't need debouncing, which causes a slight delay.
Has this changes somwhere?

tonyghiani reviewed (commented):
Overall great job Kerry, I can see all the effort you've put on this work! I couldn't find critical issues, I just left a couple of notes but in general looks good!

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Fix i18n

Kerry350 committed:
Slight tweak to validaiton in machine logic and don't reset simulation with YAML content changes

Kerry350 committed:
Debounce custom samples changes

Kerry350 committed:
Add strong typing to action metadata

flash1293 commented:
I don't have any processing steps but it still doesn't want to simulate:

<img width="710" height="338" alt="Screenshot 2025-12-09 at 10 54 26" src="https://github.com/user-attachments/assets/d3406259-9f7f-42c5-99c4-bc1655c214f7" />

A few other things (like switching to interactive) don't work, I think it's all downstream from this problem:

<img width="622" height="227" alt="Screenshot 2025-12-09 at 10 56 04" src="https://github.com/user-attachments/assets/21d7d8bd-b7e1-47f9-aea1-005bc04e8a1a" />

Kerry350 committed:
Simplify handling of identifiers (child machines maintain their notion of ID)

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Make IDs a little cheaper (we only track additive steps)

Kerry350 committed:
Strip updated_at and fix dark mode

Kerry350 committed:
Improve selector

Kerry350 committed:
Ensure there's only one unsaved changes prompt (stops history warnings)

Kerry350 committed:
Add additional mode switch case

Kerry350 committed:
CSS tweak

Kerry350 committed:
Ensure the editor itself doesn't use stale simulatin results (we keep them around for the table)

Kerry350 committed:
Better hasChanges logic

flash1293 reviewed (commented):
Tested again and it works really well now! Only issue I can still see is that typeahead doesn't really work for conditions - it doesn't suggest `field`/`eq`/`neq` etc.

I think that and field/value autocomplete are follow-ups we should still do.

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Enhance autocomplete / suggestions for conditions, fix indentation.

Kerry350 committed:
Fix property on Jest test

Kerry350 committed:
Skip focussed outline rendering when suggestions are active

Kerry350 commented:
@flash1293 Thanks Joe.

> Only issue I can still see is that typeahead doesn't really work for conditions - it doesn't suggest field/eq/neq etc.

I've added a commit here: https://github.com/elastic/kibana/pull/242743/commits/351850643fdc96a57ffcf1bec506be7c8edba818 that should vastly improve the conditions experience. You should get autocomplete for the operators now, and I've also added a bunch of snippets to help too ("Field contains...", "Field equals..." etc). Also tweaked the indentation options to make that more explicit. 

https://github.com/user-attachments/assets/2a89f94e-efc2-482a-a4e3-394ba674f660

rmyz reviewed (approved):
`src/platform/packages/shared/kbn-synthtrace/src/scenarios/sample_logs.ts` change LGTM

Kerry350 committed:
Ensure YAML mode syncs it's DSL back once set up (so that IDs are in sync)

This just ensures instant simulation rendering when switching modes

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

mdbirnstiehl commented on x-pack/platform/packages/shared/kbn-streamlang-yaml-editor/src/components/step_actions.tsx:
```diff
@@ -0,0 +1,136 @@
+    } else if (!canRunSimulation) {
+      tooltipContent = i18n.translate('xpack.streamlang.yamlEditor.stepActions.cannotRun', {
+        defaultMessage:
+          'Simulation cannot be run. Ensure the YAML is valid and changes are purely additive for partial data sources.',
```
```suggestion
          'Simulation can't run. Ensure the YAML is valid and changes for partial data sources are strictly additive.',
```

mdbirnstiehl reviewed (commented):
I had a few suggestions/comments. Let me know if anything I suggested needs to be refined or missed the point.

weltenwort reviewed (approved):
synthtrace changes LGTM

Great analysis in https://github.com/elastic/streams-program/discussions/646. That definitely helped my understanding. I see a few related function names in which `where` was not replaced with `condition`. Are these kept unchanged intentionally?

Kerry350 committed:
Property name change after merging main

Kerry350 committed:
Update x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/simulation_errors.tsx

Co-authored-by: Mike Birnstiehl <114418652+mdbirnstiehl@users.noreply.github.com>

Kerry350 committed:
Update x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/simulation_errors.tsx

Co-authored-by: Mike Birnstiehl <114418652+mdbirnstiehl@users.noreply.github.com>

Kerry350 committed:
Update x-pack/platform/packages/shared/kbn-streamlang/src/actions/action_metadata.ts

Co-authored-by: Mike Birnstiehl <114418652+mdbirnstiehl@users.noreply.github.com>

Kerry350 committed:
Update x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/edit_mode_toggle.tsx

Co-authored-by: Mike Birnstiehl <114418652+mdbirnstiehl@users.noreply.github.com>

Kerry350 committed:
Update x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/simulation_errors.tsx

Co-authored-by: Mike Birnstiehl <114418652+mdbirnstiehl@users.noreply.github.com>

Kerry350 committed:
Update x-pack/platform/packages/shared/kbn-streamlang-yaml-editor/src/components/step_actions.tsx

Co-authored-by: Mike Birnstiehl <114418652+mdbirnstiehl@users.noreply.github.com>

Kerry350 committed:
Update x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/simulation_errors.tsx

Co-authored-by: Mike Birnstiehl <114418652+mdbirnstiehl@users.noreply.github.com>

Kerry350 committed:
Update x-pack/platform/packages/shared/kbn-streamlang-yaml-editor/src/components/step_actions.tsx

Co-authored-by: Mike Birnstiehl <114418652+mdbirnstiehl@users.noreply.github.com>

Kerry350 committed:
Copy change

Kerry350 committed:
Merge branch '446-basic-yaml-editor' of github.com:Kerry350/kibana into 446-basic-yaml-editor

Kerry350 committed:
Rename functions and utils

There's no functional changes here, just DX.

Kerry350 committed:
Fix i18n

flash1293 reviewed (approved):
Wow, much much better, thanks for this fix! LGTM for now, we can improve smaller nits over time. Thanks for this massive amount of work

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

kibanamachine committed:
[Scout] Automated config manifest updates

yngrdyn reviewed (approved):
Approving regeneration of SHA in `workflows_extensions` plugin

Kerry350 committed:
Merge remote-tracking branch 'upstream/main' into 446-basic-yaml-editor

Kerry350 committed:
Merge branch '446-basic-yaml-editor' of github.com:Kerry350/kibana into 446-basic-yaml-editor
